samples/split.c

```diff
@@ -0,0 +1,182 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE
+
+*/
+
+/* The pupose of this code is to demontrate the library apis and need proper error handling and optimization */
+
+#include <getopt.h>
+#include <unistd.h>
+#include <htslib/sam.h>
+
+/// print_usage - print the demo_usage
+/** @param fp pointer to the file / terminal to which demo_usage to be dumped
+returns nothing
+*/
+static void print_usage(FILE *fp)
+{
+    fprintf(fp, "Usage: split [-i infile] [<infile>] -o outdir\n\
+Splits the input file alignments to read1 and read2 and saves as 1.sam and 2.sam in given directory\n");
+    return;
+}
+
+/// main_demo - start of the demo
+/** @param argc - count of arguments
+ *  @param argv - pointer to array of arguments
+returns 1 on failure 0 on success
+*/
+int main(int argc, char *argv[])
+{
+    static const struct option lopts[] = {
+        {NULL, 0, NULL, 0}
+    };
+    const char *inname = NULL, *outdir = NULL;
+    char *file1 = NULL, *file2 = NULL;
+    int c = 0, ret = EXIT_FAILURE, size = 0;
+    samFile *infile = NULL, *outfile1 = NULL, *outfile2 = NULL;
+    sam_hdr_t *in_samhdr = NULL;
+    bam1_t *bamdata = NULL;
+    size_t cnt = 0;
+
+    //[-i infile] [<infile>] [-o output directory]
+    while ((c = getopt_long(argc, argv, "i:o:", lopts, NULL)) >= 0)
+    {
+        switch (c)
+        {
+        case 'i':                           //infile name
+            inname = optarg;
+            break;
+        case 'o':                           //output directory
+            outdir = optarg;
+            break;
+        case '?':                           //unknown options found!
+            print_usage(stderr);
+            goto end;
+            break;
+        }
+    }
+
+    if (argc == 1 && isatty(STDIN_FILENO)) {
+        print_usage(stdout);
+        ret = EXIT_SUCCESS;
+        goto end;
+    }
+    //infile name
+    if (!inname && ((argc - optind) == 1))
+        inname = argv[optind];
+    if (!inname)
+        inname = "-";
+    //output directory
+    if (!outdir) {
+        fprintf(stderr, "Output directory not specified\n");
```

Comment:
One for @daviesrob to consider maybe, but we never use `hts_log` functions from samtools and I don't know why.  We could have instead:

```
hts_log_warning("no output directory specified");
```

or 

```
hts_log(HTS_LOG_WARNING, "split","no output directory specified");
```

I'm not sure why we always use fprintf here, but we even simulate the same style of output:

https://github.com/samtools/samtools/blob/develop/sam_view.c#L1164

I suspect it's just history.

That said, it can be useful to not start with this most demonstration programs as we don't want to bamboozle people with a large API to start with.